tests/scripts/generate_psa_tests.py

```diff
@@ -272,7 +275,15 @@ def test_cases_for_key_type_key_generation(
         """
         result = 'PSA_SUCCESS'
 
-        import_dependencies = [psa_want_symbol(kt.name)]
+        # Even if HMAC support is disabled it is possible to generate HMAC key
+        no_key_type_dependency = frozenset([
+            'PSA_KEY_TYPE_HMAC'
+        ])
+
+        if kt.name in no_key_type_dependency:
+            import_dependencies = []
+        else:
+            import_dependencies = [psa_want_symbol(kt.name)]
```

Comment:
I don't understand this patch. Why wasn't it enough to add `PSA_KEY_TYPE_HMAC` to `ALWAYS_SUPPORTED`?

Reply:
`PSA_KEY_TYPE_HMAC` has been added to `ALWAYS_SUPPORTED` and as a result HMAC generate/import tests (when !PSA_KEY_TYPE_HMAC) are removed

Here we need to remove `PSA_KEY_TYPE_HMAC` dependency from HMAC `key_generation` tests. They can pass even if `PSA_KEY_TYPE_HMAC` is not defined.

Reply:
Ok, thanks for the explanation. I still don't understand this patch however: what's special about HMAC that doesn't apply to DERIVE and RAW_DATA?

Reply:
This still doesn't answer why `DERIVE` and `RAW_DATA` didn't need such an exception. The answer to that is that it's never been an issue because they're enabled in all our testing. And that, in turn, is the case because they're enabled in all builds: they're forced on in `include/mbedtls/config_psa.h`.

Should we do the same for HMAC? Now that I think of it, I think the answer is no — and for a reason that implies we should not resolve this issue by ignoring that HMAC is disabled. The reason for having DERIVE and RAW_DATA always on is that their import and generation does not require any type-specific handling, so it would be more hassle to check for support for no real benefit. Earlier, I wrote that this applies to HMAC, but that's not quite true. For HMAC, it applies with our current implementation, but it might not apply to a different implementation, and there are benefits to doing it differently.

HMAC allows practically arbitrary long keys, but long keys are reduced (they're hashed) before doing anything else. An implementation can do it either when the key is created or at the beginning of each operation. The advantage of doing it on key creation is that it uses less RAM (and storage if the key is persistent) (and in particular the reduced key can always fit in a fixed-size buffer) and it's faster unless the key is only ever used once. The advantage of doing it at usage time is that it's slightly less complex. If we switch to hashing at key creation time, that part of the key creation process will be under `#if PSA_WANT_KEY_TYPE_HMAC`.

Will we ever switch? I'm not sure. If we switch, we'll also have to reduce keys at load time when loading from older persistent storage, which adds to the complexity. Long keys needing to be reduced are rare, so the memory benefits would not apply to most users.

Even if we don't switch, the fact that HMAC key creation works regardless of `PSA_WANT_KEY_TYPE_HMAC` is a detail of our implementation. So I'd rather not make it a property of the test script. Instead, I think we should add `#define PSA_WANT_KEY_TYPE_HMAC` in `include/mbedtls/config_psa.h` along with `DERIVE` and `RAW_DATA`, with a comment stating that _for our implementation as it is now_ HMAC key creation has no special handling and therefore is always enabled.

Reply:
My understanding was that it we want to disable `not_supported` HMAC generate/import tests when `!PSA_KEY_TYPE_HMAC ` we need to also remove this dependency from the positive test (as HMAC key creation works regardless of `PSA_WANT_KEY_TYPE_HMAC`).

> This still doesn't answer why DERIVE and RAW_DATA didn't need such an exception. The answer to that is that it's never been an issue because they're enabled in all our testing.

Yes, `PSA_WANT_KEY_TYPE_HMAC` must be defined in all configurations as otherwise `not_supported` tests would fail.

> I think we should add #define PSA_WANT_KEY_TYPE_HMAC in include/mbedtls/config_psa.h along with DERIVE and RAW_DATA, with a comment stating that for our implementation as it is now HMAC key creation has no special handling and therefore is always enabled.

If we set `PSA_WANT_KEY_TYPE_HMAC` as always enabled it seems that no change in test generator script will be needed as `not_supported` HMAC generate/import tests will be always skipped and `generate` HMAC key tests will be always executed?

Reply:
Right, if we make `PSA_WANT_KEY_TYPE_HMAC` then no change is needed in the generator script other than adding HMAC to `ALWAYS_SUPPORTED`. (Even that isn't strictly needed, but we're trying to avoid generating test cases with dependencies that cannot be met, because one day we want to enforce that all tests cases are executed on the CI in at least one configuration.)

Reply:
I filed https://github.com/Mbed-TLS/mbedtls/issues/5830 since it turns out this didn't need fixing here.